client/src/components/tables/DataTable.js

```diff
@@ -0,0 +1,206 @@
+ *
+ *
+ * @typedef {Object} buttonConfig
+ * @property {string} label                    Button inner text.
+ * @property {(row: object) => any?} callback  Callback function, passed the row that the click happened on.
+ * @property {modalComponent} modal            Modal component to bind to the button.
+ *
+ *
+ * @typedef {Object} column
+ * @property {string} id                       ID string for the column, e.g. `'start_date'`.
+ *                                             Must match a property name in the target data.
+ * @property {string} name                     Label for the column, e.g. `'Start Date'`
+ * @property {undefined|'date'|'button'} type  Type of column, used for formatting.
+ *                                             * `date` will format the column as a date, using a UTC ISO timestamp.
+ *                                             * `button` will render a button (see the `button` property).
+ *                                             * `modal` will open a model (see the `button.modal` property).
+ * @property {buttonConfig?} button            Configuration for buttons, if `type` is `'button'`.
+ *
+ *
+ * @typedef {(columns: column[] ) => Promise<Object[]> } fetchFunction
+ * @typedef {(props: modalComponentProps) => JSX.Element?} modalComponent
+ */
+
+/**
+ * Data table component, renders arbitrary columns using a provided data fetching callback.
+ * @param {object}   props
+ * @param {column[]} props.columns             Array of column configuration objects.
+ *                                             Defines the table's structure and behaviour.
+ * @param {fetchFunction} props.fetchData      Async function to fetch data for this table.
+ *                                             Must return an array of objects, with keys matching column IDs.
+ * @returns {JSX.Element}
+ */
+export const DataTable = ({ columns, fetchData, data }) => {
+  const classes = useTableStyles();
+  const [order, setOrder] = useState('asc');
+  const [isLoadingData, setLoadingData] = useState(false);
+  const [isPendingRequests, setIsPendingRequests] = useState(true);
+  const [rows, setRows] = useState([]);
+  const [orderBy, setOrderBy] = useState('name');
+
+  const handleRequestSort = (_, property) => {
+    const isAsc = orderBy === property && order === 'asc';
+    setOrder(isAsc ? 'desc' : 'asc');
+    setOrderBy(property);
+  };
+
+  const loadData = useCallback(async () => {
+    setLoadingData(true);
+    const rowsData = await fetchData(columns);
+
+    setRows(rowsData);
+    setIsPendingRequests(rowsData.length > 0);
+    setLoadingData(false);
+  }, [setRows, fetchData, columns]);
+
+  const sort = (array) => sortObjects(array, orderBy, order);
+
+  useEffect(() => {
+    if (data) {
+      setRows(data);
+    } else {
+      loadData();
+    }
+  }, [data, loadData]);
+
+  const types = Object.fromEntries(columns.map((column) => [column.id, column.type]));
```

Comment:
I think I see what's going on in these blocks but it took me a minute to read through. Some comments might help, I think especially on what we expect modalsState to do and look like

Reply:
Good call, modalState is a bit of an oversight for typing especially. I'll give it some jsdoc lines.